h2/src/main/org/h2/table/TableFilter.java

```diff
@@ -224,22 +224,24 @@ public PlanItem getBestPlanItem(Session s, int level) {
             if (select != null) {
                 sortOrder = select.getSortOrder();
             }
-            item = table.getBestPlanItem(s, masks, this, sortOrder);
+            item = table.getBestPlanItem(s, masks, filters, filter, sortOrder);
             // The more index conditions, the earlier the table.
             // This is to ensure joins without indexes run quickly:
             // x (x.a=10); y (x.b=y.b) - see issue 113
-            item.cost -= item.cost * indexConditions.size() / 100 / level;
+            item.cost -= item.cost * indexConditions.size() / 100 / (filter + 1);
         }
         if (nestedJoin != null) {
             setEvaluatable(nestedJoin);
-            item.setNestedJoinPlan(nestedJoin.getBestPlanItem(s, level));
+            item.setNestedJoinPlan(nestedJoin.getBestPlanItem(s, filters, filter));
             // TODO optimizer: calculate cost of a join: should use separate
             // expected row number and lookup cost
             item.cost += item.cost * item.getNestedJoinPlan().cost;
         }
         if (join != null) {
             setEvaluatable(join);
-            item.setJoinPlan(join.getBestPlanItem(s, level));
```

Comment:
Probably there was a bug because `join.getBestPlanItem` was called with the same level. As far as I understand it must be called with its own level.